File: config/pixel/ts/react.ts


export function createElement(component: Function, props?: Object, ...children: any[]) {
    // console.log(typeof component, typeof props, typeof children);
    const element = component(props);
    if (!element) {
        return;
    }
    if (typeof element.component === 'function') {
        return createElement(element.component, element.props, ...children);
    }
    console.log(element.component, element.props);
    if (element.props) {
        for (const prop in element.props) {
            console.log('-' + prop, element.props[prop]);
        }
    }

    console.log(`Children (${children.length}):`);
    for (let i = 0; i < children.length; i++) {
        console.log(children[i]);
    }

    return { ...element, children };
}

export function Fragment() {
    return {
        component: 'Fragment',
    };
}

File: config/pixel/ts/react.lua
local ____lualib = require("lualib_bundle")
local __TS__ObjectAssign = ____lualib.__TS__ObjectAssign
local ____exports = {}
function ____exports.createElement(self, component, props, ...)
    local children = {...}
    local element = component(nil, props)
    if not element then
        return
    end
    if type(element.component) == "function" then
        return ____exports.createElement(nil, element.component, element.props, ...)
    end
    print(element.component, element.props)
    if element.props then
        for prop in pairs(element.props) do
            print("-" .. prop, element.props[prop])
        end
    end
    print(("Children (" .. tostring(#children)) .. "):")
    do
        local i = 0
        while i < #children do
            print(children[i + 1])
            i = i + 1
        end
    end
    return __TS__ObjectAssign({}, element, {children = children})
end
function ____exports.Fragment(self)
    return {component = "Fragment"}
end
return ____exports
